File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/ObjectContextModule.vb
Public Class ObjectContextModule
    Implements IHttpModule

    Private Const sessionKey As String = "objectcontext.current"

    Public Sub Dispose() Implements IHttpModule.Dispose

    End Sub

    Public Sub Init(ByVal context As HttpApplication) Implements IHttpModule.Init
        AddHandler context.PostAcquireRequestState,
            Sub(s, e)
                If Not HttpContext.Current Is Nothing AndAlso
                   Not HttpContext.Current.Session Is Nothing Then
                    CurrentContext = New BlogModelContainer()
                End If
            End Sub

        AddHandler context.ReleaseRequestState,
            Sub(s, e)
                If Not HttpContext.Current Is Nothing AndAlso
                   Not HttpContext.Current.Session Is Nothing Then
                    CurrentContext.Dispose()
                    CurrentContext = Nothing
                End If

            End Sub
    End Sub

    Public Shared Property CurrentContext As BlogModelContainer
        Get
            Return TryCast(HttpContext.Current.Session(sessionKey), BlogModelContainer)
        End Get
        Set(ByVal value As BlogModelContainer)
            HttpContext.Current.Session(sessionKey) = value
        End Set
    End Property
End Class

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/ObjectContextModule.cs
using System.Web;
using CoolMvcBlog.Models;
using System;
namespace CoolMvcBlog.Utils
{
    public class ObjectContextModule : IHttpModule
    {
        private const string sessionKey = "objectcontext.current";

        public void Dispose()
        {
        }

        public void Init(HttpApplication context)
        {
            context.PostAcquireRequestState += (s, e) =>
                {
                    if (HttpContext.Current != null && HttpContext.Current.Session != null)
                        CurrentContext = new BlogModelContainer();
                };

            context.ReleaseRequestState += (s, e) =>
                {
                    if (HttpContext.Current != null && HttpContext.Current.Session != null)
                    {
                        CurrentContext.Dispose();
                        CurrentContext = null;
                    }
                };
        }

        public static BlogModelContainer CurrentContext
        {
            get 
            {
                return (BlogModelContainer) HttpContext.Current.Session[sessionKey];
            }
            private set 
            {
                HttpContext.Current.Session[sessionKey] = value;
            }
        }

        //private stati void checkContext()
        //{
        //    if (HttpContext.Current == null ||
        //        HttpContext.Current.Session == null)
        //    {
        //        throw new InvalidOperationException("Session not found while retrieving ObjectContext");
        //    }
        //}
    }
}